controller/src/main/java/io/pravega/controller/store/stream/records/RecordHelper.java

```diff
@@ -0,0 +1,220 @@
+                segmentsWithOffset = epochRecord.getSegments().stream().collect(Collectors.toMap(x -> x,
+                        x -> truncationRecord.getStreamCut().getOrDefault(x, 0L)));
+            } else {
+                // case 3: overlap between requested epoch and stream cut.
+                // take segments from stream cut that are from or after this epoch.
+                // take remaining segments from this epoch.
+                segmentsWithOffset = new HashMap<>();
+                // all segments from stream cut that have epoch >= this epoch
+                List<StreamSegmentRecord> fromStreamCut = truncationRecord.getSpan().entrySet().stream()
+                        .filter(x -> x.getValue() >= epochRecord.getEpoch())
+                        .map(Map.Entry::getKey)
+                        .collect(Collectors.toList());
+
+                // add segments from the truncation record with corresponding offsets
+                fromStreamCut.forEach(x -> segmentsWithOffset.put(x, truncationRecord.getStreamCut().get(x)));
+
+                // put remaining segments as those that dont overlap with ones taken from streamCut.
+                // Note: we will use the head of these segments, basically offset = 0
+                epochRecord.getSegments().stream().filter(x -> fromStreamCut.stream().noneMatch(x::overlaps))
+                        .forEach(x -> segmentsWithOffset.put(x, 0L));
+            }
+        }
+        return segmentsWithOffset;
+    }
+
+    // region scale helper methods
+    /**
+     * Method to validate supplied scale input. It performs a check that new ranges are identical to sealed ranges.
+     *
+     * @param segmentsToSeal segments to seal
+     * @param newRanges      new ranges to create
+     * @param currentEpoch   current epoch record
+     * @return true if scale input is valid, false otherwise.
+     */
+    public static boolean validateInputRange(final List<Long> segmentsToSeal,
+                                             final List<AbstractMap.SimpleEntry<Double, Double>> newRanges,
+                                             final EpochRecord currentEpoch) {
+        boolean newRangesPredicate = newRanges.stream().noneMatch(x -> x.getKey() >= x.getValue() &&
+                x.getKey() >= 0 && x.getValue() > 0);
+
+        List<AbstractMap.SimpleEntry<Double, Double>> oldRanges = segmentsToSeal.stream()
+                .map(segmentId -> currentEpoch.getSegments().stream().filter(x -> x.segmentId() == segmentId).findFirst().map(x ->
+                        new AbstractMap.SimpleEntry<>(x.getKeyStart(), x.getKeyEnd())))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toList());
+
+        return newRangesPredicate && reduce(oldRanges).equals(reduce(newRanges));
+    }
+
+    /**
+     * Method to check scale operation can be performed with given input.
```

Comment:
Can we explain in words what the condition is? This is a good way for the reviewer to verify that the code matches what the developer is trying to achieve.